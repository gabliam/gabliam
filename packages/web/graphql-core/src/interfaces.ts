import { ApolloServerPluginLandingPageGraphQLPlaygroundOptions } from 'apollo-server-core';
import { GraphQLResolveInfo } from 'graphql';

export type extractArgsFn = (
  source: any,
  args: any,
  context: any,
  info: GraphQLResolveInfo
) => any;

export interface GraphqlConfig {
  endpointUrl: string;
  playground: ApolloServerPluginLandingPageGraphQLPlaygroundOptions;
  playgroundEnabled: boolean;
}
