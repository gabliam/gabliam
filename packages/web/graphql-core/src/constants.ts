/** Type in registry */
export const TYPE = {
  Controller: 'GraphqlResolver',
};

export const GRAPHQL_CONFIG = Symbol('GABLIAM/GRAPHQL_PLUGIN_CONFIG');

export const DEBUG_PATH = 'Gabliam:Plugin:GraphqlCore';

export const DEFAULT_END_POINT_URL = '/graphql';

export const APPOLLO_SERVER_BASE = Symbol('APOLLO_SERVER_BASE');
