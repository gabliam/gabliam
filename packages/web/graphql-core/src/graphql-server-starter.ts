import {
  Container,
  Registry,
  Config,
  OnMissingBean,
  Bean,
} from '@gabliam/core';
import {
  APP,
  SERVER,
  ServerStarter,
  WebPluginConfig,
  WEB_PLUGIN_CONFIG,
  SERVER_STARTER,
} from '@gabliam/web-core';
import { ApolloServerBase } from 'apollo-server-core';
import { createServer } from 'http';

@Config(90)
export class GraphqlServerStarterConfig {
  @OnMissingBean(SERVER_STARTER)
  @Bean(SERVER_STARTER)
  createGraphqlServerStarter() {
    return new GraphqlServerStarter();
  }
}

export class GraphqlServerStarter implements ServerStarter {
  apolloServer: ApolloServerBase;

  start(container: Container, registry: Registry) {
    const restConfig = container.get<WebPluginConfig>(WEB_PLUGIN_CONFIG);
    const app = container.get<any>(APP);
    const port = restConfig.port;

    const server = createServer(<any>app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);

    this.apolloServer.installSubscriptionHandlers(server);

    container.bind(SERVER).toConstantValue(server);

    /* istanbul ignore next */
    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    /* istanbul ignore next */
    function onListening(): void {
      const addr = server.address();
      let bind = '';
      if (typeof addr === 'string') {
        bind = `pipe ${addr}`;
      } else if (addr && addr.port) {
        bind = `port ${addr.port}`;
      }
      console.log(`Listening on ${bind}`);
    }
  }
}
