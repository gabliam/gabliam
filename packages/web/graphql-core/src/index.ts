export { TYPE } from './constants';
export * from './graphql-core-plugin';
export * from './interfaces';
export * from './resolver-decorator';
export {
  Arg,
  ArgOptions,
  Args,
  ArgsType,
  Authorized,
  createParamDecorator,
  createMethodDecorator,
  Ctx,
  Directive,
  Extensions,
  registerEnumType,
  Field,
  FieldOptions,
  FieldResolver,
  Info,
  InputType,
  InputTypeOptions,
  InterfaceType,
  InterfaceTypeOptions,
  Mutation,
  ObjectType,
  ObjectTypeOptions,
  PubSub,
  Query,
  Resolver,
  Root,
  Subscription,
  SubscriptionOptions,
  createUnionType,
  UseMiddleware,
  ID, Int, ClassType, Publisher
} from 'type-graphql';
