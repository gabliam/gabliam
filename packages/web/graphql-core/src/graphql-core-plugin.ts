import { Container, GabliamPlugin, Registry, Scan } from '@gabliam/core';
import { WebConfiguration } from '@gabliam/web-core';
import { ApolloServerBase } from 'apollo-server-core';
import { GraphQLSchema } from 'graphql';
import { buildSchema } from 'type-graphql';
import { GRAPHQL_CONFIG, TYPE } from './constants';
import { GraphqlConfig } from './interfaces';

@Scan()
export abstract class GraphqlCorePlugin implements GabliamPlugin {
  async build(container: Container, registry: Registry) {
    const graphqlPluginConfig = container.get<GraphqlConfig>(GRAPHQL_CONFIG);
    const controllerIds = registry.get(TYPE.Controller).map(val => val.target);
    const schema = await buildSchema({
      resolvers: controllerIds as [Function, ...Function[]],
      container: {
        get(someClass) {
          return container.get(someClass);
        },
      },
    });

    const apolloServer = this.getApolloServer(
      container,
      registry,
      graphqlPluginConfig,
      schema
    );

    container.get(WebConfiguration).addServerConfigs(server => {
      apolloServer.installSubscriptionHandlers(server);
      return server;
    });
  }

  abstract getApolloServer(
    container: Container,
    registry: Registry,
    graphqlPluginConfig: GraphqlConfig,
    schema: GraphQLSchema
  ): ApolloServerBase;
}
