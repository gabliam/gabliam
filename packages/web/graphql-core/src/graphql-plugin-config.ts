import { Bean, Joi, PluginConfig, Value } from '@gabliam/core';
import { ApolloServerPluginLandingPageGraphQLPlaygroundOptions } from 'apollo-server-core';
import d from 'debug';
import { DEBUG_PATH, DEFAULT_END_POINT_URL, GRAPHQL_CONFIG } from './constants';
import { GraphqlConfig } from './interfaces';

const debug = d(`${DEBUG_PATH}:GraphqlPluginConfig`);

export const GraphiqlOptionsValidator = Joi.object().keys({
  endpointURL: Joi.string(),
  subscriptionsEndpoint: Joi.string(),
  query: Joi.string(),
  variables: Joi.object(),
  operationName: Joi.number(),
  result: Joi.object(),
  passHeader: Joi.string(),
});

@PluginConfig()
export class GraphqlPluginConfig {
  @Value('application.graphql.endpointUrl', Joi.string())
  private endpointUrl: string = DEFAULT_END_POINT_URL;

  @Value('application.graphql.playground.enabled', Joi.boolean())
  private playgroundEnabled = false;

  @Value('application.graphql.playground.config', GraphiqlOptionsValidator)
  private playground: ApolloServerPluginLandingPageGraphQLPlaygroundOptions;

  @Bean(GRAPHQL_CONFIG)
  creatreConfig(): GraphqlConfig {
    const graphqlConfig: GraphqlConfig = {
      endpointUrl: this.endpointUrl,
      playground: this.playground,
      playgroundEnabled: this.playgroundEnabled,
    };
    debug('GraphqlConfig', graphqlConfig);

    return graphqlConfig;
  }
}
