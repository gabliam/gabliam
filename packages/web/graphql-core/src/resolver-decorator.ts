import { injectable, Register } from '@gabliam/core';
import { ClassType, Resolver } from 'type-graphql';
import {
  AbstractClassOptions,
  ClassTypeResolver,
} from 'type-graphql/dist/decorators/types';
import { TYPE } from './constants';

export function GabResolver(): ClassDecorator;

export function GabResolver(options: AbstractClassOptions): ClassDecorator;
export function GabResolver(
  typeFunc: ClassTypeResolver,
  options?: AbstractClassOptions,
): ClassDecorator;
export function GabResolver(
  objectType: ClassType,
  options?: AbstractClassOptions,
): ClassDecorator;
export function GabResolver(
  objectTypeOrTypeFuncOrMaybeOptions?: Function | AbstractClassOptions,
  maybeOptions?: AbstractClassOptions,
): ClassDecorator {
  const r = Resolver(
    <any>objectTypeOrTypeFuncOrMaybeOptions,
    maybeOptions,
  );
  return (target: any) => {
    r(target);
    injectable()(target);
    Register({ type: TYPE.Controller, id: target })(target);
  };
}
