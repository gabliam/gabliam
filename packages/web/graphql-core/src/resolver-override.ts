import { injectable, Register } from '@gabliam/core';
import * as type from 'type-graphql';
import { TYPE } from './constants';

const oldResolver = type.Resolver as any;

(<any>type).Resolver = function(...args: any[]) {
  const r = oldResolver(...args);
  return (target: any) => {
    r(target);
    injectable()(target);
    Register({ type: TYPE.Controller, id: target })(target);
  };
};
