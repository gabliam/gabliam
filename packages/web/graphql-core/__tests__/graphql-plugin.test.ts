import { Config, Gabliam } from '@gabliam/core';
import * as sinon from 'sinon';
import { GabliamTest } from '@gabliam/core/src/testing';
import ExpressPlugin from '@gabliam/express';
import * as path from 'path';
import { GraphqlPlugin } from './graphql-plugin';

const build = sinon.spy(GraphqlPlugin.prototype, 'build');
let gab: Gabliam;

describe('with config', () => {
  beforeAll(() => {
    const g = new GabliamTest(
      new Gabliam({ scanPath: path.resolve(__dirname, './fixtures') })
        .addPlugin(ExpressPlugin)
        .addPlugin(GraphqlPlugin),
    );
    gab = g.gab;

    @Config()
    class Conf {}
    g.addClass(Conf);
  });

  afterAll(() => {
    build.resetHistory();
  });

  test('gabliam build', async () => {
    await gab.build();
    expect(build.calledOnce).toBe(true);
  });
});

describe('without config', () => {
  beforeAll(() => {
    gab = new Gabliam();
    gab.addPlugin(ExpressPlugin).addPlugin(GraphqlPlugin);
  });

  afterAll(() => {
    build.resetHistory();
  });

  test('gabliam build', async () => {
    await expect(gab.build()).rejects.toMatchSnapshot();
  });
});
