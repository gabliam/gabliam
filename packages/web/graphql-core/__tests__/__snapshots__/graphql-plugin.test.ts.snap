// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`without config gabliam build 1`] = `[Error: Empty \`resolvers\` array property found in \`buildSchema\` options!]`;
